### PR TITLE
feat(management): manage Org & Env for parameters

### PR DESCRIPTION
**Pull request by phiz71**
Closes gravitee-io/issues#4642

### CODE REVIEW

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ConfigServiceImpl.java [hunk 1]
```diff
@@ -71,9 +71,37 @@ public boolean portalLoginForced() {
...
     public PortalConfigEntity getPortalConfig() {
+        return this.getPortalConfig(GraviteeContext.getCurrentEnvironment());
+    }
+
+    @Override
+    public PortalConfigEntity getPortalConfig(String environmentId) {
         PortalConfigEntity portalConfigEntity = new PortalConfigEntity();
         Object[] objects = getObjectArray(portalConfigEntity);
 
+        loadConfigByReference(objects, portalConfigEntity, environmentId, ParameterReferenceType.ENVIRONMENT);
```
Comment (ytvnr) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ConfigServiceImpl.java:82]: maybe you can let loadConfigByReference get the object array and avoid this parameter ?

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ConfigServiceImpl.java [hunk 2]
```diff
@@ -71,9 +71,37 @@ public boolean portalLoginForced() {
...
+    public ConsoleConfigEntity getConsoleConfig() {
+        return this.getConsoleConfig(GraviteeContext.getCurrentOrganization());
+    }
+
+    @Override
+    public ConsoleConfigEntity getConsoleConfig(String organizationId) {
+        ConsoleConfigEntity consoleConfigEntity = new ConsoleConfigEntity();
+        Object[] objects = getObjectArray(consoleConfigEntity);
+
+        loadConfigByReference(objects, consoleConfigEntity, organizationId, ParameterReferenceType.ORGANIZATION);
```
Comment (ytvnr) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ConfigServiceImpl.java:98]: same remark

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ConfigServiceImpl.java [hunk 3]
```diff
@@ -71,9 +71,37 @@ public boolean portalLoginForced() {
...
+        enhanceFromConfigFile(consoleConfigEntity);
+
+        return consoleConfigEntity;
+    }
+
+    private void loadConfigByReference(Object[] objects, AbstractCommonConfigEntity configEntity, String referenceId, ParameterReferenceType referenceType) {
```
Comment (ytvnr) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ConfigServiceImpl.java:104]: from my previous remark, what do you think of
```suggestion
    private void loadConfigByReference(AbstractCommonConfigEntity configEntity, String referenceId, ParameterReferenceType referenceType) {
```

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ConfigServiceImpl.java [hunk 4]
```diff
@@ -161,40 +186,39 @@ private String getFirstValueOrDefault(final List<String> values, final String de
...
+        reCaptcha.setSiteKey(reCaptchaService.getSiteKey());
+        consoleConfigEntity.setReCaptcha(reCaptcha);
+
+        final ConsoleConfigEntity.Newsletter newsletter = new ConsoleConfigEntity.Newsletter();
         newsletter.setEnabled(newsletterService.isEnabled());
-        portalConfigEntity.setNewsletter(newsletter);
+        consoleConfigEntity.setNewsletter(newsletter);
+    }
+
+    private void enhanceAbstractFromConfigFile(AbstractCommonConfigEntity abstractCommonConfigEntity) {
```
Comment (ytvnr) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ConfigServiceImpl.java:206]: useless ?

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/EnvironmentServiceImpl.java [hunk 5]
```diff
@@ -15,24 +15,27 @@
...
+import io.gravitee.rest.api.service.exceptions.EnvironmentNotFoundException;
+import io.gravitee.rest.api.service.exceptions.TechnicalManagementException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
```
Comment (ytvnr) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/EnvironmentServiceImpl.java:38]: only reformat ? :D

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -524,6 +525,10 @@ public UserEntity create(NewExternalUserEntity newExternalUserEntity, boolean ad
 
     private UserEntity create(NewExternalUserEntity newExternalUserEntity, boolean addDefaultRole, boolean autoRegistrationEnabled) {
         try {
+            /*
+             TODO: getCurrentEnvironnemenet and call database to fetch the corresponding organization OR add parameters in methods
```
Comment (ytvnr) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/UserServiceImpl.java:529]: real TODO or you forgot to delete it ? :)

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/configuration/application/ApplicationTypeServiceImpl.java [hunk 7]
```diff
@@ -17,18 +17,14 @@
 
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
-import io.gravitee.repository.exceptions.TechnicalException;
-import io.gravitee.rest.api.model.PortalConfigEntity;
+import io.gravitee.rest.api.model.parameters.PortalConfigEntity;
```
Comment (ytvnr) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/configuration/application/ApplicationTypeServiceImpl.java:20]: reformat ? :D

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/spring/GraviteeJavaMailSenderImpl.java [hunk 8]
```diff
@@ -46,61 +48,85 @@ public GraviteeJavaMailSenderImpl(ParameterService parameterService, EventManage
...
-            initialized = true;
+            this.mailSenderByReference.put(ref, mailSender);
         }
 
-        return super.getSession();
+        return mailSender.getSession();
+    }
+
+    @NotNull
+    private ReferenceContext getReferenceContext() {
```
Comment (ytvnr) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/spring/GraviteeJavaMailSenderImpl.java:72]: I would have prefered to use `GraviteeContext.getCurrentContext();` see next comment

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/spring/GraviteeJavaMailSenderImpl.java [hunk 9]
```diff
@@ -114,4 +140,28 @@ private Properties loadProperties() {
 
         return properties;
     }
+
+    class ReferenceContext {
+        private String referenceId;
+        private ParameterReferenceType referenceType;
```
Comment (ytvnr) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/spring/GraviteeJavaMailSenderImpl.java:146]: Can you enrich `GraviteeContext.getCurrentContext()` ?
to have a String field and an enum field for referenceType ?

i think it is sad to dupplicate the work here and I would prefer a more complete object in GraviteeContext.

wdyt ?

**Review by ytvnr — COMMENTED** [review]
great job !
some remarks and suggestions

File: gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/organization/ConsoleResource.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import javax.ws.rs.container.ResourceContext;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+
+import static io.gravitee.rest.api.model.permissions.RolePermissionAction.*;
+
+/**
+ * Defines the REST resources to manage Portal.
+ *
+ * @author Nicolas GERAUD (nicolas.geraud at graviteesource.com)
```
[the remark this answers is not in the document]
Comment (phiz71) [reply gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/organization/ConsoleResource.java:43]: 😇

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/UserServiceImpl.java (hunk 6 above)
Comment (phiz71) [reply gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/UserServiceImpl.java:529]: a TODO not related to this feature

**Review by ytvnr — APPROVED** [review]

**Pull request merged**